appinventor/components/src/com/google/appinventor/components/runtime/CSVFile.java

```diff
@@ -0,0 +1,239 @@
+
+                        // Open asset file
+                        final InputStream inputStream = form.openAsset(filename);
+
+                        AsynchUtil.runAsynchronously(new Runnable() {
+                            @Override
+                            public void run() {
+                                readCSV(inputStream);
+                            }
+                        });
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                } else {
+                    form.dispatchPermissionDeniedEvent(CSVFile.this, "ReadFrom", permission);
+                }
+            }
+        });
+    }
+
+    private void readCSV(InputStream inputStream) {
+        readingDone = false;
+        try {
+            // TODO: Taken form File.java. To be replaced to reduce redundancy.
+            InputStreamReader input = new InputStreamReader(inputStream);
+            StringWriter output = new StringWriter();
+            char [] buffer = new char[4096];
+            int offset = 0;
+            int length = 0;
+            while ((length = input.read(buffer, offset, 4096)) > 0) {
+                output.write(buffer, 0, length);
+            }
+
+            final String result = output.toString().replaceAll("\r\n", "\n");
+
+            // Parse rows from the result
+            rows = CsvUtil.fromCsvTable(result);
+
+            // Construct column lists from rows
+            constructColumnsFromRows();
+
+            // TODO: Notify data reading done (for async race condition)
+            readingDone = true;
+
+            for (int i = 0; i < dataComponents.size(); ++i) {
+                ChartDataBase dataComponent = dataComponents.get(i);
+                YailList columns = dataComponentColumns.get(i);
+
+                dataComponent.importFromCSVAsync(this, columns);
+            }
+
+            dataComponents.clear();
+            dataComponentColumns.clear();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Rows property getter method
+     *
+     * @return a YailList representing the list of strings to be picked from
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR)
+    public YailList Rows() {
+        return rows;
+    }
+
+
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR)
+    public YailList Columns() {
+        return columns;
+    }
+
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR)
+    public YailList ColumnNames() {
+        return columnNames;
+    }
+
+    /**
+     * Sets the source file to parse CSV from.
+     *
+     * @param source  Source file name
+     */
+    @SimpleProperty(category = PropertyCategory.BEHAVIOR)
+    @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_ASSET)
+    public void SourceFile(String source) {
```

Comment:
Careful handling in SourceFile setter needs to be done, since reading might still be occurring while the user tries to change the source file.

Reply:
Synchronized methods are now used. To be tested.